### PR TITLE
scrape: Speed up scrape once we have reached sample_limit

### LINKED ISSUE #5690: CPU on 100% load when job sample_limit is breached
## Bug Report
**What did you do?**
running prometheus with `sample_limit` configured for jobs

**What did you expect to see?**
business as usual if `sample_limit` is breached

**What did you see instead? Under which circumstances?**
if `sample_limit` is breached all CPU cores are 100% loaded

**Environment**

* System information:

	Linux 4.15.0-1040-aws x86_64

* Prometheus version:

prometheus, version 2.10.0 (branch: HEAD, revision: d20e84d0fb64aff2f62a977adc8cfb656da4e286)
  build user:       root@a49185acd9b0
  build date:       20190525-12:28:13
  go version:       go1.12.5

* Alertmanager version:

	not related

* Prometheus configuration file:
```
- job_name: failingjob
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  sample_limit: 5000
  static_configs:
  - targets:
    - 10.50.2.118:8294
    - 10.50.1.179:8294
    - 10.50.5.219:8294
    - 10.50.0.191:8294
    - 10.50.1.93:8294
    - 10.50.1.56:8294

```

* Alertmanager configuration file:
```
insert configuration here (if relevant to the issue)
```

level=warn ts=2019-06-20T07:02:19.971Z caller=scrape.go:937 component="scrape manager" scrape_pool=failingjob target=http://10.50.1.56:8294/metrics msg="append failed" err="sample limit exceeded

### PR DESCRIPTION
**Pull request by roidelapluie**
Fix #5690

As we reach the sample_limit, it is no longer needed to parse metrics
check cache & do relabelling -- we should just count the number of metrics left.

### CODE REVIEW

File: scrape/scrape.go [hunk 1]
```diff
@@ -1134,7 +1137,9 @@ loop:
 			}
 
 			var ref uint64
-			ref, err = app.Add(lset, t, v)
+			if err != errSampleLimit {
```
Comment (brian-brazil) [inline scrape/scrape.go]: Shouldn't we be checking that errSampleLimit is not nil?

A comment here would also be useful.

**Comment by roidelapluie** [comment]
@brian-brazil I think this one is better

File: scrape/scrape.go (hunk 1 above)
Comment (roidelapluie) [reply scrape/scrape.go]: I've put a second test that will be more effective and only do something if we breach the limit.

**Comment by roidelapluie** [comment]
still need test

File: scrape/scrape.go [hunk 2]
```diff
@@ -1108,11 +1108,20 @@ loop:
 				targetScrapeSampleOutOfBounds.Inc()
 				continue
 			case errSampleLimit:
-				// Keep on parsing output if we hit the limit, so we report the correct
-				// total number of samples scraped.
 				sampleLimitErr = err
 				added++
-				continue
+
+				// Count the remaining metrics in place, to avoid touching the
```
Comment (brian-brazil) [inline scrape/scrape.go]: They're samples, not metrics.

File: scrape/scrape.go [hunk 3]
```diff
@@ -1108,11 +1108,20 @@ loop:
...
+				// storage layer.
+				var remaining int
+				added, err = countRemainingMetrics(p)
+				// Cout of the remaining metrics could return an error. If
```
Comment (brian-brazil) [inline scrape/scrape.go]: Count

This isn't C++ stream IO :)

**Comment by roidelapluie** [comment]
ready for review :)

**Comment by roidelapluie** [comment]
Note that it will no longer convert the samples to series and therefore the count will also contain the series that are in in droppedSeries.

As we are in a case of breaking sample limits anyway, I think it is acceptable as we want to quickly finish the scrape.

**Comment by roidelapluie** [comment]
I tested manually with a 5.000.000 /metrics endpoint with a limit at 5000





scrape time goes to 5.5s to 0.8s

**Comment by roidelapluie** [comment]
Benchmark

```
benchmark                                              old ns/op       new ns/op       delta
BenchmarkScrapeLoopAppendSampleLimit/100x1.1-8         43556           13052           -70.03%
BenchmarkScrapeLoopAppendSampleLimit/100x1.5-8         59586           17147           -71.22%
BenchmarkScrapeLoopAppendSampleLimit/100x2-8           77491           23140           -70.14%
BenchmarkScrapeLoopAppendSampleLimit/100x5-8           200492          53887           -73.12%
BenchmarkScrapeLoopAppendSampleLimit/100x10-8          416163          107292          -74.22%
BenchmarkScrapeLoopAppendSampleLimit/100x100-8         3866957         1026895         -73.44%
BenchmarkScrapeLoopAppendSampleLimit/100x1000-8        37042364        10557452        -71.50%
BenchmarkScrapeLoopAppendSampleLimit/100x10000-8       355821480       105005649       -70.49%
BenchmarkScrapeLoopAppendSampleLimit/1000x1.1-8        455672          118611          -73.97%
BenchmarkScrapeLoopAppendSampleLimit/1000x1.5-8        629227          162510          -74.17%
BenchmarkScrapeLoopAppendSampleLimit/1000x2-8          777553          219727          -71.74%
BenchmarkScrapeLoopAppendSampleLimit/1000x5-8          1869539         534025          -71.44%
BenchmarkScrapeLoopAppendSampleLimit/1000x10-8         4112363         1062212         -74.17%
BenchmarkScrapeLoopAppendSampleLimit/1000x100-8        38682654        10414268        -73.08%
BenchmarkScrapeLoopAppendSampleLimit/1000x1000-8       373562383       102323049       -72.61%
BenchmarkScrapeLoopAppendSampleLimit/1000x10000-8      3488088420      1053417567      -69.80%
BenchmarkScrapeLoopAppendSampleLimit/10000x1.1-8       4451121         1169922         -73.72%
BenchmarkScrapeLoopAppendSampleLimit/10000x1.5-8       5887881         1658950         -71.82%
BenchmarkScrapeLoopAppendSampleLimit/10000x2-8         8155631         2182022         -73.25%
BenchmarkScrapeLoopAppendSampleLimit/10000x5-8         20445595        5400332         -73.59%
BenchmarkScrapeLoopAppendSampleLimit/10000x10-8        39684982        10706167        -73.02%
BenchmarkScrapeLoopAppendSampleLimit/10000x100-8       357654689       104841981       -70.69%
BenchmarkScrapeLoopAppendSampleLimit/10000x1000-8      3656158681      1094190744      -70.07%
BenchmarkScrapeLoopAppendSampleLimit/10000x10000-8     34971398423     10529485385     -69.89%

benchmark                                              old allocs     new allocs     delta
BenchmarkScrapeLoopAppendSampleLimit/100x1.1-8         452            8              -98.23%
BenchmarkScrapeLoopAppendSampleLimit/100x1.5-8         612            8              -98.69%
BenchmarkScrapeLoopAppendSampleLimit/100x2-8           812            8              -99.01%
BenchmarkScrapeLoopAppendSampleLimit/100x5-8           2012           8              -99.60%
BenchmarkScrapeLoopAppendSampleLimit/100x10-8          4012           8              -99.80%
BenchmarkScrapeLoopAppendSampleLimit/100x100-8         40012          8              -99.98%
BenchmarkScrapeLoopAppendSampleLimit/100x1000-8        400012         12             -100.00%
BenchmarkScrapeLoopAppendSampleLimit/100x10000-8       4000015        42             -100.00%
BenchmarkScrapeLoopAppendSampleLimit/1000x1.1-8        4412           8              -99.82%
BenchmarkScrapeLoopAppendSampleLimit/1000x1.5-8        6012           8              -99.87%
BenchmarkScrapeLoopAppendSampleLimit/1000x2-8          8012           8              -99.90%
BenchmarkScrapeLoopAppendSampleLimit/1000x5-8          20012          9              -99.96%
BenchmarkScrapeLoopAppendSampleLimit/1000x10-8         40012          11             -99.97%
BenchmarkScrapeLoopAppendSampleLimit/1000x100-8        400012         48             -99.99%
BenchmarkScrapeLoopAppendSampleLimit/1000x1000-8       4000016        409            -99.99%
BenchmarkScrapeLoopAppendSampleLimit/1000x10000-8      40000024       4020           -99.99%
BenchmarkScrapeLoopAppendSampleLimit/10000x1.1-8       44012          48             -99.89%
BenchmarkScrapeLoopAppendSampleLimit/10000x1.5-8       60012          60             -99.90%
BenchmarkScrapeLoopAppendSampleLimit/10000x2-8         80012          81             -99.90%
BenchmarkScrapeLoopAppendSampleLimit/10000x5-8         200012         185            -99.91%
BenchmarkScrapeLoopAppendSampleLimit/10000x10-8        400013         368            -99.91%
BenchmarkScrapeLoopAppendSampleLimit/10000x100-8       4000019        4010           -99.90%
BenchmarkScrapeLoopAppendSampleLimit/10000x1000-8      40000034       40029          -99.90%
BenchmarkScrapeLoopAppendSampleLimit/10000x10000-8     400000033      40029          -99.99%

benchmark                                              old bytes       new bytes     delta
BenchmarkScrapeLoopAppendSampleLimit/100x1.1-8         11864           320           -97.30%
BenchmarkScrapeLoopAppendSampleLimit/100x1.5-8         16024           320           -98.00%
BenchmarkScrapeLoopAppendSampleLimit/100x2-8           21224           320           -98.49%
BenchmarkScrapeLoopAppendSampleLimit/100x5-8           52426           321           -99.39%
BenchmarkScrapeLoopAppendSampleLimit/100x10-8          104428          322           -99.69%
BenchmarkScrapeLoopAppendSampleLimit/100x100-8         1040473         338           -99.97%
BenchmarkScrapeLoopAppendSampleLimit/100x1000-8        10400851        532           -99.99%
BenchmarkScrapeLoopAppendSampleLimit/100x10000-8       104004181       2093          -100.00%
BenchmarkScrapeLoopAppendSampleLimit/1000x1.1-8        114858          340           -99.70%
BenchmarkScrapeLoopAppendSampleLimit/1000x1.5-8        156470          347           -99.78%
BenchmarkScrapeLoopAppendSampleLimit/1000x2-8          208488          360           -99.83%
BenchmarkScrapeLoopAppendSampleLimit/1000x5-8          520596          409           -99.92%
BenchmarkScrapeLoopAppendSampleLimit/1000x10-8         1040759         493           -99.95%
BenchmarkScrapeLoopAppendSampleLimit/1000x100-8        10403471        2286          -99.98%
BenchmarkScrapeLoopAppendSampleLimit/1000x1000-8       104031461       19987         -99.98%
BenchmarkScrapeLoopAppendSampleLimit/1000x10000-8      1040093232      196992        -99.98%
BenchmarkScrapeLoopAppendSampleLimit/10000x1.1-8       1151819         3377          -99.71%
BenchmarkScrapeLoopAppendSampleLimit/10000x1.5-8       1570982         4295          -99.73%
BenchmarkScrapeLoopAppendSampleLimit/10000x2-8         2094094         5906          -99.72%
BenchmarkScrapeLoopAppendSampleLimit/10000x5-8         5235436         13800         -99.74%
BenchmarkScrapeLoopAppendSampleLimit/10000x10-8        10471692        27646         -99.74%
BenchmarkScrapeLoopAppendSampleLimit/10000x100-8       104664960       303643        -99.71%
BenchmarkScrapeLoopAppendSampleLimit/10000x1000-8      1041993888      3033552       -99.71%
BenchmarkScrapeLoopAppendSampleLimit/10000x10000-8     10401993808     3033536       -99.97%
```

**Comment by brian-brazil** [comment]
> Note that it will no longer convert the samples to series and therefore the count will also contain the series that are in in droppedSeries.

That's not okay, the scrape_samples_post_metric_relabeling needs to still have the right value. (And add unittests if we don't already have them for that).

**Comment by brian-brazil** [comment]
> Note that it will no longer convert the samples to series and therefore the count will also contain the series that are in in droppedSeries.

That's not okay, scrape_samples_post_metric_relabeling needs to have the right value and we should have tests verifying that.

**Comment by roidelapluie** [comment]
ack, will see to add that test

**Pull request closed without merging**